File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.swift
import Foundation

let nmk = readLine()!.split{$0==" "}.map{Int($0)!}
let (N, M, K) = (nmk[0], nmk[1], nmk[2])
var li = (0..<N).map { _ in Int(readLine()!)! }
var st = Array(repeating: 0, count: 4*N)

func make(_ l: Int, _ r: Int, _ i: Int) -> Int {
    if l == r {
        st[i] = li[l]
        return st[i]
    }
    let m = (l+r)/2
    st[i] = make(l, m, i*2) + make(m+1, r, i*2+1)
    return st[i]
}
func update(_ l: Int, _ r: Int, _ i: Int, _ x: Int, _ diff: Int) {
    if x < l || x > r {
        return
    }
    st[i] += diff
    if l == r { return }
    let m = (l+r)/2
    update(l, m, i*2, x, diff)
    update(m+1, r, i*2+1, x, diff)
}
func sum(_ l: Int, _ r: Int, _ i: Int, _ x: Int, _ y: Int) -> Int {
    if r < x || l > y {
        return 0
    }
    if l >= x && r <= y {
        return st[i]
    }
    let m = (l+r)/2
    return sum(l, m, i*2, x, y) + sum(m+1, r, i*2+1, x, y)
}
_=make(0, li.count-1, 1)
for _ in 0..<M+K {
    let abc = readLine()!.split{$0==" "}.map{Int($0)!}
    let (a,b,c) = (abc[0], abc[1]-1, abc[2])
    if a == 1 {
        update(0, li.count-1, 1, b, c-li[b])
        li[b] = c
    } else {
        print(sum(0, li.count-1, 1, b, c-1))
    }
}

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

// 	메모리 109148KB	시간 508ms
public class Main {
    static long[] tree;

    static void init(long[] li) {
        int treeSize = li.length + 1;
        tree = new long[treeSize+1];

        for (int i = 1; i <= li.length; i++) {
            int j = i;

            while (j <= treeSize) {
                tree[j] += li[i-1];
                j += (j & -j);
            }
        }
    }
    static void update(int index, long diff) {
        int j = index;

        while (j <= tree.length) {
            tree[j] += diff;
            j += (j & -j);
        }
    }
    static long sum(int to) {
        int i = to;
        long ret = 0;

        while (i != 0) {
            ret += tree[i];
            i -= (i & -i);
        }

        return ret;
    }
    static long subSum(int from, int to) {
        return sum(to) - sum(from-1);
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        int SUBSUM = 2;
        StringBuilder sb = new StringBuilder();

        long[] li = new long[N];
        for (int i = 0; i < N; i++) {
            li[i] = Long.parseLong(br.readLine());
        }
        init(li);
        for (int i = 0; i < M + K; i++) {
            st = new StringTokenizer(br.readLine());
            long command = Integer.parseInt(st.nextToken());
            int from = Integer.parseInt(st.nextToken());
            long to = Long.parseLong(st.nextToken());

            if (command == SUBSUM) {
                sb.append(subSum(from, (int)to)).append("\n");
            } else {
                update(from, to - li[from-1]);
                li[from-1] = to;
            }
        }
        System.out.println(sb);
    }
}